src/storage/txn/scheduler.rs

```diff
@@ -800,46 +834,71 @@ impl Scheduler {
-        SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(cid), "snapshot"]).inc();
+    fn get_snapshot(&mut self, ctx: &Context, cids: Vec<u64>) {
+        for cid in &cids {
+            SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(*cid), "snapshot"]).inc();
+        }
+        let cids1 = cids.clone();
+        let cids2 = cids.clone();
         let ch = self.schedch.clone();
-        let cb = box move |(cb_ctx, snapshot)| {
+        let cb = box move |(cb_ctx, snapshot): (CbContext, EngineResult<Box<Snapshot>>)| {
             match ch.send(Msg::SnapshotFinished {
-                cid: cid,
+                cids: cids1,
                 cb_ctx: cb_ctx,
                 snapshot: snapshot,
             }) {
                 Ok(_) => {}
                 e @ Err(TransportError::Closed) => info!("channel closed, err {:?}", e),
-                Err(e) => panic!("send SnapshotFinish failed cmd id {}, err {:?}", cid, e),
+                Err(e) => {
+                    panic!("send SnapshotFinish failed cmd ids {:?}, err {:?}",
+                           cids2,
+                           e)
+                }
             }
         };
 
-        if let Err(e) = self.engine.async_snapshot(self.extract_context(cid), cb) {
-            SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(cid), "async_snap_err"])
-                .inc();
-            self.finish_with_err(cid, Error::from(e));
+        if let Err(e) = self.engine.async_snapshot(ctx, cb) {
+            for cid in cids {
+                SCHED_STAGE_COUNTER_VEC
+                    .with_label_values(&[self.get_ctx_tag(cid), "async_snap_err"])
+                    .inc();
+
+                let e = e.maybe_clone().unwrap_or_else(|| {
+                    error!("async snapshot failed for cid={}, error {:?}", cid, e);
+                    EngineError::Other(box_err!("{:?}", e))
+                });
+                self.finish_with_err(cid, Error::from(e));
+            }
         }
     }
 
     /// Event handler for the completion of get snapshot.
     ///
     /// Delivers the command along with the snapshot to a worker thread to execute.
     fn on_snapshot_finished(&mut self,
-                            cid: u64,
+                            cids: Vec<u64>,
                             cb_ctx: CbContext,
                             snapshot: EngineResult<Box<Snapshot>>) {
-        debug!("receive snapshot finish msg for cid={}", cid);
-        match snapshot {
-            Ok(snapshot) => {
-                SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(cid), "snapshot_ok"])
-                    .inc();
-                self.process_by_worker(cid, cb_ctx, snapshot);
-            }
-            Err(e) => {
-                SCHED_STAGE_COUNTER_VEC.with_label_values(&[self.get_ctx_tag(cid), "snapshot_err"])
-                    .inc();
-                self.finish_with_err(cid, Error::from(e));
+        for cid in cids {
```

Comment:
Yes, we need to call `finish_with_err` for every cid, then there would be another for loop in Err(_). Or make `finish_with_err` takes a `&[cid]`, so that the for loop is moved out, I think that might be overkill.